File: screenrecorder/src/main/java/com/sunplus/screenrecorder/media/glutils/EGLBase.java
package com.sunplus.screenrecorder.media.glutils;

/**
 * Created by w.feng on 2018/10/11
 * Email: fengweisb@gmail.com
 */
public abstract class EGLBase {
  public static final Object EGL_LOCK = new Object();
  public static final int EGL_RECORDABLE_ANDROID = 0x3142;
  public static final int EGL_CONTEXT_CLIENT_VERSION = 0x3098;
  public static final int EGL_OPENGL_ES2_BIT = 4;
  public static final int EGL_OPENGL_ES3_BIT_KHR = 0x0040;

  public static EGLBase createFrom(final IContext sharedContext,
                                   final boolean withDepthBuffer, final boolean isRecordable) {

    return createFrom(3, sharedContext, withDepthBuffer, 0, isRecordable);
  }

  public static EGLBase createFrom(final IContext sharedContext,
                                   final boolean withDepthBuffer, final int stencilBits,
                                   final boolean isRecordable) {

    return createFrom(3, sharedContext,
        withDepthBuffer, stencilBits, isRecordable);
  }

  public static EGLBase createFrom(final int maxClientVersion,
                                   final IContext sharedContext, final boolean withDepthBuffer,
                                   final int stencilBits, final boolean isRecordable) {

    if (isEGL14Supported() && ((sharedContext == null)
        || (sharedContext instanceof EGLBaseHigh.Context))) {

      return new EGLBaseHigh(maxClientVersion,
          (EGLBaseHigh.Context) sharedContext,
          withDepthBuffer, stencilBits, isRecordable);
    } else {
      return new EGLBaseLow(maxClientVersion,
          (EGLBaseLow.Context) sharedContext,
          withDepthBuffer, stencilBits, isRecordable);
    }
  }

  public static abstract class IContext {
    public abstract long getNativeHandle();

    public abstract Object getEGLContext();
  }

  public static abstract class IConfig {
  }

  public interface IEglSurface {
    public void makeCurrent();

    public void swap();

    public IContext getContext();

    /**
     * swap with presentation time[ns]
     * only works well now when using EGLBase14
     */
    public void swap(final long presentationTimeNs);

    public void release();

    public boolean isValid();
  }

  public static boolean isEGL14Supported() {
    return true;
  }

  public abstract void release();

  public abstract String queryString(final int what);

  public abstract int getGlVersion();

  public abstract IContext getContext();

  public abstract IConfig getConfig();

  public abstract IEglSurface createFromSurface(final Object nativeWindow);

  public abstract IEglSurface createOffscreen(final int width, final int height);

  public abstract void makeDefault();

  public abstract void sync();
}

File: screenrecorder/src/main/java/com/almosting/screenrecorder/media/glutils/EGLBase.kt
package com.almosting.screenrecorder.media.glutils

import com.almosting.screenrecorder.media.glutils.EGLBaseHigh.Context

/**
 * Created by w.feng on 2018/10/11
 * Email: fengweisb@gmail.com
 */
abstract class EGLBase {
  abstract class IContext {
    abstract fun getNativeHandle(): Long
    abstract fun getEGLContext(): Any?
  }

  abstract class IConfig
  interface IEglSurface {
    fun makeCurrent()
    fun swap()
    fun getContext(): IContext?

    /**
     * swap with presentation time[ns]
     * only works well now when using EGLBase14
     */
    fun swap(presentationTimeNs: Long)
    fun release()
    fun isValid(): Boolean
  }

  abstract fun release()
  abstract fun queryString(what: Int): String?
  abstract fun getGlVersion(): Int
  abstract fun getContext(): IContext?
  abstract fun getConfig(): IConfig?
  abstract fun createFromSurface(nativeWindow: Any?): IEglSurface?
  abstract fun createOffscreen(width: Int, height: Int): IEglSurface?
  abstract fun makeDefault()
  abstract fun sync()

  companion object {
    val EGL_LOCK: Any? = Any()
    const val EGL_RECORDABLE_ANDROID = 0x3142
    const val EGL_CONTEXT_CLIENT_VERSION = 0x3098
    const val EGL_OPENGL_ES2_BIT = 4
    const val EGL_OPENGL_ES3_BIT_KHR = 0x0040
    fun createFrom(
      sharedContext: IContext?,
      withDepthBuffer: Boolean, isRecordable: Boolean
    ): EGLBase? {
      return createFrom(3, sharedContext, withDepthBuffer, 0, isRecordable)
    }

    fun createFrom(
      sharedContext: IContext?,
      withDepthBuffer: Boolean, stencilBits: Int,
      isRecordable: Boolean
    ): EGLBase? {
      return createFrom(
        3, sharedContext,
        withDepthBuffer, stencilBits, isRecordable
      )
    }

    fun createFrom(
      maxClientVersion: Int,
      sharedContext: IContext?, withDepthBuffer: Boolean,
      stencilBits: Int, isRecordable: Boolean
    ): EGLBase? {
      return if (isEGL14Supported() && (sharedContext == null
            || sharedContext is Context)
      ) {
        EGLBaseHigh(
          maxClientVersion,
          sharedContext as Context?,
          withDepthBuffer, stencilBits, isRecordable
        )
      } else {
        EGLBaseLow(
          maxClientVersion,
          sharedContext as EGLBaseLow.Context?,
          withDepthBuffer, stencilBits, isRecordable
        )
      }
    }

    fun isEGL14Supported(): Boolean {
      return true
    }
  }
}